File: frontend/src/functions/adapter-http/impl/author.ts
import { HttpInstance, Response } from "../server"
import { IdResponse, LimitAndOffsetFilter, Link, ListResult, mapFromOrderList, OrderList } from "./generic"
import { SimpleAnnotation } from "./annotations"
import { SourceMappingMetaItem, SourceMappingMetaItemForm } from "./source-tag-mapping"
import { AlreadyExists, NotFound, ResourceNotExist, ResourceNotSuitable } from "../exception"

export function createAuthorEndpoint(http: HttpInstance): AuthorEndpoint {
    return {
        list: http.createQueryRequest("/api/authors", "GET", {
            parseQuery: mapFromAuthorFilter
        }),
        create: http.createDataRequest("/api/authors", "POST"),
        get: http.createPathRequest(id => `/api/authors/${id}`),
        update: http.createPathDataRequest(id => `/api/authors/${id}`, "PATCH"),
        delete: http.createPathRequest(id => `/api/authors/${id}`, "DELETE")
    }
}

function mapFromAuthorFilter(data: AuthorFilter): any {
    return {
        ...data,
        order: mapFromOrderList(data.order),
        annotationIds: data.annotationIds?.length ? data.annotationIds.join(",") : undefined
    }
}

/**
 * 作者。
 */
export interface AuthorEndpoint {
    /**
     * 查询作者列表。
     */
    list(filter: AuthorFilter): Promise<Response<ListResult<Author>>>
    /**
     * 新建作者。
     * @exception ALREADY_EXISTS ("Author", "name", name) 作者重名
     * @exception NOT_EXISTS ("annotations", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("annotations", id) 指定的资源不适用。对于annotations，此注解的target要求不能应用于此种类的tag
     */
    create(form: AuthorCreateForm): Promise<Response<IdResponse, AuthorExceptions["create"]>>
    /**
     * 查看作者。
     * @exception NOT_FOUND
     */
    get(id: number): Promise<Response<DetailAuthor, NotFound>>
    /**
     * 更改作者。
     * @exception NOT_FOUND
     * @exception ALREADY_EXISTS ("Author", "name", name) 作者重名
     * @exception NOT_EXISTS ("annotations", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("annotations", id) 指定的资源不适用。对于annotations，此注解的target要求不能应用于此种类的tag
     */
    update(id: number, form: AuthorUpdateForm): Promise<Response<null, AuthorExceptions["update"]>>
    /**
     * 删除作者。
     * @exception NOT_FOUND
     */
    delete(id: number): Promise<Response<null, NotFound>>
}

export interface AuthorExceptions {
    "create": AlreadyExists<"Author", "name", string> | ResourceNotExist<"annotations", number[]> | ResourceNotSuitable<"annotations", number[]> | ResourceNotExist<"source", string>
    "update": NotFound | AlreadyExists<"Author", "name", string> | ResourceNotExist<"annotations", number[]> | ResourceNotSuitable<"annotations", number[]> | ResourceNotExist<"source", string>
}

export type AuthorType = "UNKNOWN" | "ARTIST" | "STUDIO" | "PUBLISH"

export interface Author {
    /**
     * author id。
     */
    id: number
    /**
     * 作者名称。需要遵守tag name规范。
     */
    name: string
    /**
     * 其他名称。需要遵守tag name规范。
     */
    otherNames: string[]
    /**
     * 关键字。需要遵守tag name规范。
     */
    keywords: string[]
    /**
     * 作者类型。
     */
    type: AuthorType
    /**
     * 标记为收藏。
     */
    favorite: boolean
    /**
     * 注解。
     */
    annotations: SimpleAnnotation[]
    /**
     * 评分。
     */
    score: number | null
    /**
     * 关联的项目数量。
     */
    count: number
    /**
     * 此author的颜色。
     */
    color: string | null
}

export interface DetailAuthor extends Author {
    /**
     * 简介。
     */
    description: string
    /**
     * 相关链接。
     */
    links: Link[]
    /**
     * 映射到此元数据的来源标签。
     */
    mappingSourceTags: SourceMappingMetaItem[]
}

export interface SimpleAuthor {
    id: number
    name: string
    type: AuthorType
    color: string | null
}

export interface DepsAuthor extends SimpleAuthor {
    isExported: boolean
}

export interface AuthorCreateForm {
    name: string
    otherNames?: string[] | null
    type?: AuthorType
    description?: string
    keywords?: string[]
    links?: Link[] | null
    annotations?: (string | number)[] | null
    favorite?: boolean
    score?: number | null
    mappingSourceTags?: SourceMappingMetaItemForm[] | null
}

export interface AuthorUpdateForm {
    name?: string
    otherNames?: string[] | null
    type?: AuthorType
    description?: string
    keywords?: string[]
    links?: Link[] | null
    annotations?: (string | number)[] | null
    favorite?: boolean
    score?: number | null
    mappingSourceTags?: SourceMappingMetaItemForm[] | null
}

export type AuthorFilter = AuthorQueryFilter & LimitAndOffsetFilter

export interface AuthorQueryFilter {
    query?: string
    order?: OrderList<"id" | "name" | "score" | "count" | "createTime" | "updateTime">
    type?: AuthorType
    favorite?: boolean
    annotationIds?: number[]
}

File: server/src/main/kotlin/com/heerkirov/hedge/server/dto/Author.kt
package com.heerkirov.hedge.server.dto

import com.heerkirov.hedge.server.library.form.*
import com.heerkirov.hedge.server.model.meta.Author
import com.heerkirov.hedge.server.utils.types.Opt
import com.heerkirov.hedge.server.utils.types.OrderItem

data class AuthorRes(val id: Int, val name: String, val otherNames: List<String>, val keywords: List<String>,
                     val type: Author.Type, val favorite: Boolean,
                     val annotations: List<Author.CachedAnnotation>,
                     val score: Int?, val count: Int, val color: String?)

data class AuthorSimpleRes(val id: Int, val name: String, val type: Author.Type, val isExported: Boolean, val color: String?)

data class AuthorDetailRes(val id: Int, val name: String, val otherNames: List<String>, val keywords: List<String>, val description: String,
                           val type: Author.Type, val favorite: Boolean,
                           val annotations: List<Author.CachedAnnotation>,
                           val links: List<Author.Link>,
                           val score: Int?, val count: Int, val color: String?,
                           val mappingSourceTags: List<SourceMappingMetaItem>)

data class AuthorFilter(@Limit val limit: Int,
                        @Offset val offset: Int,
                        @Search val search: String?,
                        @Order(options = ["id", "name", "score", "count", "createTime", "updateTime"])
                        val order: List<OrderItem>? = null,
                        val type: Author.Type? = null,
                        val favorite: Boolean? = null,
                        val annotationIds: List<Int>? = null)

data class AuthorCreateForm(@NotBlank val name: String,
                            val otherNames: List<String>? = null,
                            val type: Author.Type = Author.Type.UNKNOWN,
                            val keywords: List<String>? = null,
                            val description: String = "",
                            val links: List<Author.Link>? = null,
                            val annotations: List<Any>? = null,
                            val favorite: Boolean = false,
                            val score: Int? = null,
                            val mappingSourceTags: List<SourceMappingMetaItem>? = null)

data class AuthorUpdateForm(@NotBlank val name: Opt<String>,
                            val otherNames: Opt<List<String>?>,
                            val type: Opt<Author.Type>,
                            val keywords: Opt<List<String>?>,
                            val description: Opt<String>,
                            val links: Opt<List<Author.Link>?>,
                            val annotations: Opt<List<Any>?>,
                            val favorite: Opt<Boolean>,
                            val score: Opt<Int?>,
                            val mappingSourceTags: Opt<List<SourceMappingMetaItem>?>)

fun newAuthorRes(author: Author, colors: Map<Author.Type, String>) = AuthorRes(author.id, author.name,
    author.otherNames, author.keywords, author.type, author.favorite,
    author.cachedAnnotations ?: emptyList(), author.score, author.cachedCount, colors[author.type])

fun newAuthorDetailRes(author: Author, colors: Map<Author.Type, String>, mappingSourceTags: List<SourceMappingMetaItem>) = AuthorDetailRes(
    author.id, author.name, author.otherNames, author.keywords, author.description, author.type, author.favorite,
    author.cachedAnnotations ?: emptyList(), author.links ?: emptyList(),
    author.score, author.cachedCount, colors[author.type], mappingSourceTags)